Introduce `bdk_core`

evanlinjin opened the pull request:
Based on #1568
Closes #1543

### Description

Introduce `bdk_core` crate. Move types over from `bdk_chain`. Chain sources (`bdk_electrum`, `bdk_esplora` and `bdk_bitcoind_rpc`) now only depend on `bdk_core`.

### Notes to the reviewers

Please review commit-by-commit. I've moved things over, but slight API changes were necessary (mentioned in the commit messages).

### Changelog notice

* Add `bdk_core` crate which contains core types that were previously in `bdk_chain`. Including: `BlockId`, `ConfirmationBlockTime`, `CheckPoint`, `CheckPointIter`, `tx_graph::Update` and `spk_client`-types.
* Change chain sources (`bdk_esplora`, `bdk_electrum` and `bdk_bitcoind_rpc`) to only depend on `bdk_core`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

evanlinjin committed:
feat!: introduce `bdk_core`

This is an initial version with `chain_data` types ported over.

Types ported over include `BlockId`, `ConfirmationBlockTime`. The impls
for `Anchor` and `AnchorFromBlockPosition` of these types are moved to
where the traits are defined.

evanlinjin committed:
refactor(chain): change `CheckPoint::apply_changeset`

to be a separate function: `apply_changeset_to_checkpoint`.

evanlinjin committed:
feat!: move `CheckPoint` to `bdk_core`

Also add `CheckPoint::eq_ptr` which compares the internal pointers of
two `CheckPoint`s. This is used as an optimisation when comparing two
chains.

evanlinjin committed:
feat!: move `tx_graph::Update` to `bdk_core`

evanlinjin committed:
feat!: move `spk_client`s to `bdk_core`

Also introduced extension trait for builder methods that take in a
`KeychainTxOutIndex`.

`Indexed<T>` and `KeychainIndexed<T>` are also moved to `bdk_core`.

evanlinjin committed:
feat(electrum)!: depend on `bdk_core` instead of `bdk_chain`

`.populate_tx_cache` has been changed to take in a collection of
`Arc<Transaction>`.

evanlinjin committed:
feat(esplora)!: depend on `bdk_core` instead of `bdk_chain`

Helper methods have been changed to use concrete types (instead of
`A: Anchor` generic) - since `Anchor` is still part of `bdk_chain`.

Also fix esplora docs.

evanlinjin committed:
feat(bitcoind_rpc)!: depend on `bdk_core` instead of `bdk_chain`

notmandatory commented:
Please keep track if any types or functions were modified other than just because things were moved around.  Even though there are a lot of changes here if no types or logic were changed it shouldn't be too hard to review.

notmandatory commented on crates/core/src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+/// Core structures for [`TxGraph`].
+///
+/// [`TxGraph`]: https://docs.rs/bdk_chain/latest/bdk_chain/tx_graph/struct.TxGraph.html
+pub mod tx_graph {
```
Is there a reason to put `tx_graph` here instead of it's own file?

evanlinjin replied:
No specific reason. It's just for convenience and there is not much to the `tx_graph::Update` declaration anyway. Is there any downsides to this arrangement?

notmandatory replied:
No particular downside I can see, just didn't know if there was some advantage I wasn't aware of.

notmandatory reviewed (approved):
ACK dafb9aaef7d8d4bc83c8ba0845323b92b82fbb3d

Great to see this PR only moves code around and decouple the chain clients from the chain module; plus some small docs and helper function cleanup. Your commits were concise and well documented which helped a lot with review. 

I reran examples as a simple smoke test with no problems.

LLFourn commented on crates/core/src/lib.rs:
```diff
@@ -0,0 +1,113 @@
+/// Core structures for [`TxGraph`].
+///
+/// [`TxGraph`]: https://docs.rs/bdk_chain/latest/bdk_chain/tx_graph/struct.TxGraph.html
+pub mod tx_graph {
```
We shouldn't mention `TxGraph` here too much. Now this thing lives here the `Update` can be used by any other software to process transaction updates. I'll push a commit.

LLFourn commented on crates/core/src/spk_client.rs:
```diff
@@ -345,9 +323,11 @@ impl<I> SyncRequest<I> {
+    /// [`TxGraph`](../../bdk_chain/tx_graph/struct.TxGraph.html).
+    pub graph_update: crate::tx_graph::Update<A>,
+    /// The update to apply to the receiving
+    /// [`LocalChain`](../../bdk_chain/local_chain/struct.LocalChain.html).
```
Heh I guess this link is one way to do it but I think it's a mistake to mention `LocalChain` and `TxGraph` here. We are decoupling these things. What if we make a new chain type wouldn't a CheckPoint also be a valid update for it? (I'm pushing a commit).

LLFourn commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -189,9 +185,10 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+    /// [`bdk_chain`]: ../bdk_chain/index.html
+    /// [`CalculateFeeError::MissingTxOut`]: ../bdk_chain/tx_graph/enum.CalculateFeeError.html#variant.MissingTxOut
+    /// [`Wallet.calculate_fee`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee
+    /// [`Wallet.calculate_fee_rate`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee_rate
```
:pick: I'm against these kinds of doc links because they go stale without us noticing. Also they potentially link to incompatible versions. What if we change that method name etc?

It's enough to explain the concept.

Here are the remaining ones after a5d076f215cd91173f55bda0d1cc59b9dde75511:

```
crates/core/src/spk_client.rs:113:    /// This is used to update [`LocalChain`](../../bdk_chain/local_chain/struct.LocalChain.html).
crates/core/src/spk_client.rs:124:    /// [`KeychainTxOutIndex`](../../bdk_chain/indexer/keychain_txout/struct.KeychainTxOutIndex.html).
crates/core/src/spk_client.rs:358:    /// This is used to update [`LocalChain`](../../bdk_chain/local_chain/struct.LocalChain.html).
crates/electrum/src/bdk_electrum_client.rs:119:    /// [`bdk_chain`]: ../bdk_chain/index.html
crates/electrum/src/bdk_electrum_client.rs:120:    /// [`CalculateFeeError::MissingTxOut`]: ../bdk_chain/tx_graph/enum.CalculateFeeError.html#variant.MissingTxOut
crates/electrum/src/bdk_electrum_client.rs:121:    /// [`Wallet.calculate_fee`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee
crates/electrum/src/bdk_electrum_client.rs:122:    /// [`Wallet.calculate_fee_rate`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee_rate
crates/electrum/src/bdk_electrum_client.rs:188:    /// [`bdk_chain`]: ../bdk_chain/index.html
crates/electrum/src/bdk_electrum_client.rs:189:    /// [`CalculateFeeError::MissingTxOut`]: ../bdk_chain/tx_graph/enum.CalculateFeeError.html#variant.MissingTxOut
crates/electrum/src/bdk_electrum_client.rs:190:    /// [`Wallet.calculate_fee`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee
crates/electrum/src/bdk_electrum_client.rs:191:    /// [`Wallet.calculate_fee_rate`]: ../bdk_wallet/struct.Wallet.html#method.calculate_fee_rate
```

LLFourn committed:
chore(core)!: s/tx_graph::Update/TxUpdate/

We shouldn't refer to `tx_graph` in `bdk_core`. TxGraph happens to
consume `Update` but it doesn't conceptually own the definition of an
update to transactions.

I tried to also remove a lot of references to "graph" where they
shouldn't be. "graph" is a very general kind of data structure so we
should avoid referring to it where it's not relevant. `tx_update` is
much better than `graph_update`.

LLFourn reviewed (approved):
I added a a5d076f215cd91173f55bda0d1cc59b9dde75511: 

> We shouldn't refer to `tx_graph` in `bdk_core`. TxGraph happens to consume `Update` but it doesn't conceptually own the definition of an update to transactions.
> I tried to also remove a lot of references to "graph" where they shouldn't be. "graph" is a very general kind of data structure so we should avoid referring to it where it's not relevant. `tx_update` is much better than `graph_update`.

Self-ACK: a5d076f215cd91173f55bda0d1cc59b9dde75511

evanlinjin reviewed (commented):
ACK a5d076f215cd91173f55bda0d1cc59b9dde75511

`TxUpdate` makes sense. The doc changes make things clearer and easier to maintain.

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -39,14 +37,11 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
-            .full_txs()
-            .map(|tx_node| (tx_node.txid, tx_node.tx));
-
+    pub fn populate_tx_cache(&self, txs: impl IntoIterator<Item = impl Into<Arc<Transaction>>>) {
```
Ideally this function could accept an iterator of `(Txid, Arc<Transaction>)` tuples

LLFourn replied:
This function is called at startup time so computing txid here is not really worth optimizing out.
